File: Data Structures & Algorithms/intersection-of-two-linked-lists/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {

        ListNode currB = headB;

        Set<ListNode> hmB = new HashSet<>();

        while(currB != null) {
            hmB.add(currB);
            currB = currB.next;
        }
        

        ListNode currA = headA;
        while(currA != null) {
            if(hmB.contains(currA))
                return currA;
            currA = currA.next;
        }
        return null;
    }
}

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} list1
     * @param {ListNode} list2
     * @return {ListNode}
     */
    mergeTwoLists(list1, list2) {
        if (!list1) return list2;
        if (!list2) return list1;

        let curr = null;
        if(list1.val < list2.val) {
            curr = list1;
            list1 = list1.next;
        } else {
            curr = list2;
            list2 = list2.next;
        }

        let start = curr;
        while (list1 && list2) {
            if(list1.val < list2.val){
                curr.next = list1;
                list1 = list1.next;
            } else {
                curr.next = list2;
                list2 = list2.next;
            }
            curr = curr.next;
        }
        curr.next = list1 || list2;
        return start;

    }
}
